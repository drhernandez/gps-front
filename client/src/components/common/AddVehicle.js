import React from "react";
import {
  Row,
  Col,
  FormInput,
  FormFeedback,
  FormGroup,
  FormSelect,
  Form,
  Alert
} from "shards-react";
import "../../styles/addVehicle.css"
import Button from "./Button"
import { VehiclesService } from "../../api/services";
import { to } from "await-to-js";
import validations from "../../utils/ValidationsUtil";
import constants from "../../utils/Constants";
import _ from "lodash"

const errorsDefault = {
  brand: {
    required: false
  },
  brandline: {
    required: false
  },
  plate: {
    required: false
  }
}

export default class AddVehicle extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      errors: errorsDefault,
      showSppiner: false,
      showAlert: false,
      userId: props.userId,
      brands: [],
      brandLines: [],
      brand: null,
      brandLine: null,
      plate: null
    }

    this.getBrands.bind(this);
    this.getModelsByBrand.bind(this);
    this.saveVehicle.bind(this);
    this.handleBrandSelection.bind(this);
    this.handleBrandLineSelection.bind(this);
    this.invalidateError.bind(this);
  }

  async componentDidMount() {
    const brands = await this.getBrands();
    this.setState({
      brands: brands
    })
  }

  async getBrands() {
    const [err, brands] = await to(VehiclesService.getBrands());
    return err ? [] : brands;
  }

  async getModelsByBrand(brandId) {
    const [err, brandLines] = await to(VehiclesService.getBrandLines(brandId));
    return err ? [] : brandLines
  }

  async saveVehicle(e) {
    e.preventDefault();
    e.persist();
    const errors = _.clone(errorsDefault);
    errors.brand.required = !validations.validateRequired(this.state.brand);
    errors.brandline.required = !validations.validateRequired(this.state.brandLine);
    errors.plate.required = !validations.validateRequired(e.target.plate.value);

    if (Object.values(errors).find(fieldValidations => Object.values(fieldValidations).find(value => value))) {  // Si alguno de los errores está en true...
      this.setState({
        errors: errors
      })
    } else {

      this.setState({
        showSppiner: true
      })
      const vehicle = {
        user_id: this.state.userId,
        brand: this.state.brand,
        brand_line: this.state.brandLine,
        plate: e.target.plate.value
      }

      const [err, result] = await to(VehiclesService.createVehicle(vehicle));
      if (err) {
        this.setState({
          showSppiner: false,
          showAlert: true
        })
      }
      else {
        this.setState({
          showSppiner: false
        })
        this.props.addVehicle(vehicle);
      }
    }
  }

  async handleBrandSelection(e) {
    e.preventDefault();
    this.invalidateError("brand");
    const brandId = e.target.value;
    const brand = this.state.brands.find(brand => brand.id == brandId);
    const brandlines = await this.getModelsByBrand(brand.id);
    this.setState({
      brand: brand.name,
      brandLines: brandlines
    });
  }

  handleBrandLineSelection(e) {
    e.preventDefault();
    this.invalidateError("brandline");
    const brandLineId = e.target.value;
    const brandLine = this.state.brandLines.find(brandLine => brandLine.id == brandLineId);
    this.setState({
      brandLine: brandLine.name
    });
  }

  invalidateError(field) {
    let errors = this.state.errors;
    Object.keys(errors[field]).forEach(validationType => errors[field][validationType] = false)
    this.setState({
      errors: errors
    })
  }

  render() {
    return (
      <Form className="w-100 main-container p-2" onSubmit={(e) => this.saveVehicle(e)} noValidate>
        <FormGroup className="m-0">
          <Row form>
            {/* Brand */}
            <Col md="4" className="form-group">
              <label htmlFor="brand">Marca</label>
              <FormSelect 
                id="brand" 
                onChange={(e) => this.handleBrandSelection(e)}
                invalid={this.state.errors.brand.required}
              >
                <option key="0"> Seleccione una opción... </option>
                {
                  this.state.brands.map((brand) => {
                    return (<option key={brand.id} value={brand.id}>{brand.name}</option>)
                  })
                }
              </FormSelect>
              <FormFeedback>Campo requerido</FormFeedback>
            </Col>
            {/* Brand Lines */}
            <Col md="4" className="form-group">
              <label htmlFor="brandline">Categoría</label>
              <FormSelect 
                id="brandline" 
                onChange={(e) => this.handleBrandLineSelection(e)}
                invalid={this.state.errors.brandline.required}
              >
                <option key="0"> Seleccione una opción... </option>
                {
                  this.state.brandLines.map((brandline) => {
                    return (<option key={brandline.id} value={brandline.id}>{brandline.name}</option>)
                  })
                }
              </FormSelect>
              <FormFeedback>Campo requerido</FormFeedback>
            </Col>
            {/* Plate */}
            <Col md="4" className="form-group">
              <label htmlFor="plate">Patente</label>
              <FormInput
                id="plate"
                placeholder="Patente"
                invalid={this.state.errors.plate.required}
                onChange={() => this.invalidateError("plate")}
              />
              <FormFeedback>Campo requerido</FormFeedback>
            </Col>
          </Row>
        </FormGroup>
        <FormGroup className="mb-0">
          <Button type="submit" theme="accent" className="d-table ml-auto" label="Guardar vehículo" showSppiner={this.state.showSppiner}></Button>
        </FormGroup>
        {
          this.state.showAlert &&
          <FormGroup className="pt-3 mb-0">
            <Alert className="my-auto w-100" theme={constants.Themes.ERROR}>
              Ocurrió un error al intentar guardar el vehículo. Intentalo de nuevo!
          </Alert>
          </FormGroup>
        }
      </Form>
    )
  }
}