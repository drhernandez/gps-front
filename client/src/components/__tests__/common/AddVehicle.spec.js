import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

jest.mock("../../../api/services/vehiclesService", () => ({ getBrands: jest.fn(), getBrandLines: jest.fn(), createVehicle: jest.fn()}))
import VehiclesService from "../../../api/services/vehiclesService"

import AddVehicle from '../../common/AddVehicle'

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

describe('AddVehicle component', () => {

  beforeEach(async () => {
    jest.resetModules();
    
    await VehiclesService.getBrands.mockImplementation(() => {
      return Promise.resolve([
        { "id": 5, "name": "BMW" }
      ]);
    })
    await VehiclesService.getBrandLines.mockImplementation(() => {
      return Promise.resolve([
        { "id": 26, "name": "SERIE1" },
        { "id": 27, "name": "SERIE3" },
        { "id": 28, "name": "SERIE5" },
        { "id": 29, "name": "SERIE6" }
      ]);
    })
    await VehiclesService.createVehicle.mockImplementation(() => {
      return Promise.resolve({});
    })
  })

  it('snapshot test', () => {
    const component = renderer.create(
      <AddVehicle />
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should load brand on componentDidmount', async () => {
    const wrapper = shallow(
      <AddVehicle />
    )

    await flushPromises();
    wrapper.update();

    expect(wrapper.state().brands).toHaveLength(1)
    expect(wrapper.state().brands[0].name).toBe('BMW')
  })

  it('should load brandlines when a brand is selected', async () => {
    const wrapper = shallow(
      <AddVehicle />
    )

    await flushPromises();

    expect(wrapper.state().brand).toBe(null)
    expect(wrapper.state().brandLines).toHaveLength(0)

    const mockedEvent = {
      target: {
        value: 5
      },
      preventDefault: () => {}
    }
    wrapper.find('FormSelect').at(0).props().onChange(mockedEvent);

    await flushPromises();
    
    expect(wrapper.state().brand).toBe('BMW')
    expect(wrapper.state().brandLines).toHaveLength(4)
  })

  it('should set brandline on state after one is selected', async () => {
    const wrapper = shallow(
      <AddVehicle />
    )

    await flushPromises();

    wrapper.setState({
      brandLines: [
        { "id": 26, "name": "SERIE1" },
        { "id": 27, "name": "SERIE3" },
        { "id": 28, "name": "SERIE5" },
        { "id": 29, "name": "SERIE6" }
      ]
    })

    const mockedEvent = {
      target: {
        value: 27
      },
      preventDefault: () => { }
    }
    wrapper.find('FormSelect').at(1).props().onChange(mockedEvent);

    expect(wrapper.state().brandLine).toBe('SERIE3')
  })

  it('should call VehiclesService.createVehicle on form submit', async () => {
    debugger
    const parentAddVehicle = jest.fn()
    const props = {
      userId: 1,
      addVehicle: parentAddVehicle
    }
    const wrapper = shallow(
      <AddVehicle {...props}/>
    )

    await flushPromises();

    wrapper.setState({
      brand: "BMW",
      brandLine: "SERIE3"
    })

    const mockedEvent = {
      target: {
        plate: {
          value: 'AA111AA'
        }
      },
      preventDefault: () => {},
      persist: () => {}
    }
    wrapper.find('Form').props().onSubmit(mockedEvent);

    await flushPromises();

    expect(parentAddVehicle).toHaveBeenCalled()
    expect(VehiclesService.createVehicle).toHaveBeenCalledWith({
      user_id: 1,
      brand: 'BMW',
      brand_line: 'SERIE3',
      plate: 'AA111AA'
    })
  })
})