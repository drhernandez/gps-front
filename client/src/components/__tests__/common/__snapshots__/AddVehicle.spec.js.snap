// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddVehicle component snapshot test 1`] = `
<form
  className="w-100 main-container p-2"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="m-0 form-group"
  >
    <div
      className="form-row"
    >
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="brand"
        >
          Marca
        </label>
        <select
          className="form-control custom-select"
          id="brand"
          onChange={[Function]}
        >
          <option>
             Seleccione una opción... 
          </option>
        </select>
        <div
          className="invalid-feedback"
        >
          Campo requerido
        </div>
      </div>
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="brandline"
        >
          Categoría
        </label>
        <select
          className="form-control custom-select"
          id="brandline"
          onChange={[Function]}
        >
          <option>
             Seleccione una opción... 
          </option>
        </select>
        <div
          className="invalid-feedback"
        >
          Campo requerido
        </div>
      </div>
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="plate"
        >
          Patente
        </label>
        <input
          className="form-control"
          id="plate"
          onChange={[Function]}
          placeholder="Patente"
        />
        <div
          className="invalid-feedback"
        >
          Campo requerido
        </div>
      </div>
    </div>
  </div>
  <div
    className="mb-0 form-group"
  >
    <button
      className="d-table ml-auto btn btn-accent"
      onClick={[Function]}
      type="submit"
    >
      Guardar vehículo
    </button>
  </div>
</form>
`;
